File: 101.对称二叉树.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {

    /*
     * 两两出队的队列
     */
    public boolean isSymmetric(TreeNode root) {
        Queue<TreeNode> q = new LinkedList<>();
        TreeNode l = root.left;
        TreeNode r = root.right;
        if (root == null || (l==null && r ==null)){
            return true;
        }
        q.offer(l);
        q.offer(r);
        while (!q.isEmpty()){
            l=q.poll();
            r=q.poll();
            if (l==null && r==null){
                continue;
            }
            if ((l==null ||r==null)|| (l.val != r.val)){
                return false;
            }

                q.offer(l.left);
                q.offer(r.right);
                
                q.offer(l.right);
                q.offer(r.left);
            }

            // 低配的写法
            // if (( (l==null && r!=null)||(l!=null && r==null) )|| (l!=null &&r!=null)&&(l.val!= r.val)){
            //     return false ;
            // }else if (l== null &&r==null) {
            //     // 注意不是break，continue 是跳出当前循环，队列不为空就一直两两出队。
            //     continue;
            // }else{
            //     q.offer(l.left);
            //     q.offer(r.right);
                
            //     q.offer(l.right);
            //     q.offer(r.left);
            // }
        return true;
    }
    


     // 递归
    public boolean isSymmetric(TreeNode root) {
       if (root ==null){
           return true;
       }
       return deepCheck(root.left,root.right);
   }


   boolean deepCheck (TreeNode left ,TreeNode right ){
       if (left ==null && right == null){
           return true;
       }
       // if( (left == null||right == null)||(left.val!=right.val)){
       //     return false ;
       // }
       if(left ==null || right == null){
           return false;
       }
       if(left.val!=right.val){
           return false ;
       }
       return deepCheck(left.left,right.right) && deepCheck(left.right, right.left);
   }
}

File: 101.对称二叉树.js
/*
 * @lc app=leetcode.cn id=101 lang=javascript
 *
 * [101] 对称二叉树
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
 var isSymmetric = function(root) {
    if (root===null) return true
    return deepcheck(root.left,root.right)
};
function deepcheck(left,right){
    if(left===null&&right===null) return true
    if((left===null||right===null) || (left.val!==right.val))return false
    return deepcheck(left.left,right.right)&&deepcheck(left.right,right.left)  
}
// @lc code=end

